plutus-core/testlib/PlutusIR/Generators/QuickCheck/GenerateTerms.hs

```diff
@@ -0,0 +1,853 @@
+              -> GenTm (Term TyName Name DefaultUni DefaultFun ())
+genTermOfType ty = snd <$> genTerm (Just ty)
+
+-- | Generate a term, if the first argument is Nothing then we get something of any type
+-- and if the first argument is `Just ty` we get something of type `ty`.
+--
+-- Requires the type to be of kind *.
+genTerm :: Maybe (Type TyName DefaultUni ())
+        -> GenTm (Type TyName DefaultUni (), Term TyName Name DefaultUni DefaultFun ())
+genTerm mty = checkInvariants $ do
+  customF <- asks geCustomFreq
+  customG <- asks geCustomGen
+  vars <- asks geTerms
+  esc <- asks geEscaping
+  -- Prefer to generate things that bind variables until we have "enough" (20...)
+  let (letF, lamF, varAppF) = if Map.size vars < 20
+                              then (30, 50, 10)
+                              else (10, 30, 40)
+      atomic
+          | Just ty <- mty = (ty,) <$> genAtomicTerm ty
+          | otherwise      = do
+              ty <- genType $ Type ()
+              (ty,) <$> genAtomicTerm ty
+  ifAstSizeZero atomic $
+    frequency $
+      [ (10, atomic) ]                                             ++
+      [ (letF, genLet mty) ]                                       ++
+      [ (30, genForall x k a) | Just (TyForall _ x k a) <- [mty] ] ++
+      [ (lamF, genLam a b)    | Just (a, b) <- [funTypeView mty] ] ++
+      [ (varAppF, genVarApp mty) ]                                 ++
+      [ (10, genApp mty) ]                                         ++
+      [ (1, genError mty) ]                                        ++
+      [ (10, genConst mty)    | canConst mty ]                     ++
+      [ (10, genDatLet mty)   | YesEscape <- [esc] ]               ++
+      [ (10, genIfTrace)      | isNothing mty ]                    ++
+      [ (customF, customG mty) ]
+  where
+
+    checkInvariants gen = do
+      (ty, tm) <- gen
+      debug    <- asks geDebug
+      tyctx    <- asks geTypes
+      tmctx    <- asks geTerms
+      if debug then
+        case typeCheckTermInContext tyctx tmctx tm ty of
+          Left err ->
+             (error . show $ "genTerm - checkInvariants: term " <> prettyPirReadable tm
+                           <> " does not type check at type " <> prettyPirReadable ty
+                           <> " in type context " <> prettyPirReadable tyctx
+                           <> " and term context " <> prettyPirReadable tmctx
+                           <> " with error message " <> fromString err)
+          _ -> return (ty, tm)
+      else
+        return (ty, tm)
+
+    funTypeView Nothing                             = Just (Nothing, Nothing)
+    funTypeView (Just (normalizeTy -> TyFun _ a b)) = Just (Just a, Just b)
+    funTypeView _                                   = Nothing
+
+    -- Generate builtin ifthenelse and trace calls
+    genIfTrace = do
+      fun <- liftGen $ elements [IfThenElse, Trace]
+      pure (typeOfBuiltinFunction def fun, Builtin () fun)
```

Comment:
Probably because these were the ones we cared about controlling the distriubtion for when we were doing our testing. It's not impossible that the right thing to do would be to have one frequency per builtin.